ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/privacy/PrivateTransactionHandler.java

```diff
@@ -95,19 +161,29 @@ private SendRequest createSendRequest(final PrivateTransaction privateTransactio
         privateFor);
   }
 
-  private Transaction createPrivacyMarkerTransactionWithNonce(
-      final String transactionEnclaveKey,
-      final PrivateTransaction privateTransaction,
-      final Long nonce) {
-
-    return Transaction.builder()
-        .nonce(nonce)
-        .gasPrice(privateTransaction.getGasPrice())
-        .gasLimit(privateTransaction.getGasLimit())
-        .to(privacyPrecompileAddress)
-        .value(privateTransaction.getValue())
-        .payload(BytesValue.wrap(transactionEnclaveKey.getBytes(Charsets.UTF_8)))
-        .sender(privateTransaction.getSender())
-        .signAndBuild(nodeKeyPair);
+  private long getSenderNonce(
+      final PrivateTransaction privateTransaction, final String privacyGroupId) {
+    return privateStateStorage
+        .getPrivateAccountState(BytesValue.fromHexString(privacyGroupId))
+        .map(
+            lastRootHash ->
+                privateWorldStateArchive
+                    .getMutable(lastRootHash)
+                    .map(
+                        worldState -> {
+                          final Account maybePrivateSender =
+                              worldState.get(privateTransaction.getSender());
+
+                          if (maybePrivateSender != null) {
+                            return maybePrivateSender.getNonce();
+                          }
+                          return Account.DEFAULT_NONCE;
+                        })
+                    .orElse(Account.DEFAULT_NONCE))
+        .orElseGet(
+            () -> {
+              LOG.trace("Failed to find private state for privacy group id {}", privacyGroupId);
+              return Account.DEFAULT_NONCE;
```

Comment:
If there is a specific reason to return a default value instead of failing, please add it a comment.